File: leet_DP_977_squares_of_sorted_arr__DP.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums)
    {
        int n = nums.size() - 1;
        int R = n;
        int L = 0;
        vector<int> dp(n + 1);

        while (L <= R)
        {
            if (abs(nums[L]) < abs(nums[R]))
                dp[n] = nums[R] * nums[R--];
            else
                dp[n] = nums[L] * nums[L++];
            --n;
        }
        return dp;
    }
};

File: leet_DP_977_squares_of_sorted_arr__DP.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        dp = [0] * n
        R = n - 1
        L = 0
        while L <= R:
            n -= 1
            if abs(nums[L]) < abs(nums[R]):
                dp[n] = nums[R] ** 2
                R -= 1
            else:
                dp[n] = nums[L] ** 2
                L += 1
        return dp
